src/connectivity.jl

```diff
@@ -219,31 +219,32 @@ julia> strongly_connected_components(g)
 
 function strongly_connected_components end
 # see https://github.com/mauro3/SimpleTraits.jl/issues/47#issuecomment-327880153 for syntax
-@traitfn function strongly_connected_components(g::AG::IsDirected) where {T<:Integer, AG <: AbstractGraph{T}}
+empty_graph_data(type,g::AG) where {T, AG <: AbstractGraph{T}} = Dict{T,type}()
+empty_graph_data(type,g::AG) where {T<:Integer, AG <: AbstractGraph{T}} = zeros(type,nv(g))
+is_unvisited(data::AbstractVector,v::Integer) = iszero(data[v])
+is_unvisited(data::Dict,v) = !haskey(data,v)
+
+@traitfn function strongly_connected_components(g::AG::IsDirected) where {T, AG <: AbstractGraph{T}}
     zero_t = zero(T)
-    one_t = one(T)
     nvg = nv(g)
-    count = one_t
-    
-    
-    index = zeros(T, nvg)         # first time in which vertex is discovered
-    stack = Vector{T}()           # stores vertices which have been discovered and not yet assigned to any component
-    onstack = zeros(Bool, nvg)    # false if a vertex is waiting in the stack to receive a component assignment
-    lowlink = zeros(T, nvg)       # lowest index vertex that it can reach through back edge (index array not vertex id number)
-    parents = zeros(T, nvg)       # parent of every vertex in dfs
+    count = nvg  # (Counting downwards) Visitation order for the branch being explored. Backtracks when we pop an scc.
+    component_count = 1  # Index of the current component being discovered.
+    # Invariant 1: count is always smaller than component_count.
+    # Invariant 2: if rindex[v] < component_count, then v is in components[rindex[v]].
+    # This trivially lets us tell if a node belongs to a previously discovered scc without any extra bits, just inequalities.
+
+    component_root = empty_graph_data(Bool,g)    
```

Comment:
```suggestion
    is_component_root = empty_graph_data(Bool,g)    
```
a `Dict` of type `Dict{T, Bool}` can usually be replaced by a `Set{T}`. In this case, you could also think about using a `Vector{Bool}` or a `BitVector}` although with these you will waste some space in case we only have a few strongly connected components.

Reply:
Right, it's using a Vector{Bool} for now whenever T is an Integer. I'm planning to do some benchmarking with bitvector once I have a decent solution to make the loop nonquadratic.

The main issue with using anything that doesn't give you a pointer to a bool is that the boolean gets manipulated in a tight loop which is fast because of branch prediction/speculative execution. My guess is that it's not possible to insert anything complex in the Else branch of that for loop without a large performance cost.